File: house-robber/eunhwa99.java
// 시간 복잡도: O(n) - dp 배열을 한 번 순회
// 공간 복잡도: O(n) - dp 배열
class Solution {
    public int rob(int[] nums) {
       int[][] dp = new int[nums.length][2]; // 0: not robbed, 1: robbed

        if(nums.length == 0) return 0;
        if(nums.length == 1) return nums[0];

        dp[0][0] = 0;
        dp[0][1] = nums[0];

        for(int i = 1; i < nums.length; i++){
            dp[i][0] = Math.max(dp[i-1][0], dp[i-1][1]);
            dp[i][1] = dp[i-1][0] + nums[i];
        }

        return Math.max(dp[nums.length-1][0], dp[nums.length-1][1]);
    }
}

File: house-robber-ii/eunhwa99.kt
// https://leetcode.com/problems/house-robber-ii/description/
class Solution {
    fun rob(nums: IntArray): Int {
        if (nums.size == 1) return nums[0]
        return maxOf(rob(nums, 0, nums.size - 2), rob(nums, 1, nums.size - 1))
    }

    private fun rob(nums: IntArray, start: Int, end: Int): Int {
        val dp = IntArray(nums.size) { 0 }
        dp[start] = nums[start]
        dp[start + 1] = maxOf(nums[start], nums[start + 1])
        for (i in start + 2..end) {
            dp[i] = maxOf(dp[i - 1], dp[i - 2] + nums[i])
        }
        return dp[end]
    }
}
// TC: O(n)
// SC: O(n)
